### PR TITLE
allow for `u8` to be used as hold/freeze reason

### PR DESCRIPTION
**Pull request by kianenigma**
..without needing to provide your own `newtype` around it.  

This will allow `type Reason = u8` to be used as `FreezeReason` and `HoldReason`, which I think is a nice simplification if one doens't want to deal with the complications. 

At the same time, it is a bit of an anti-pattern.

Putting it out there to check people's vibes.

### CODE REVIEW

File: substrate/frame/support/src/traits/misc.rs [hunk 1]
```diff
@@ -48,6 +48,10 @@ impl VariantCount for () {
 	const VARIANT_COUNT: u32 = 0;
 }
 
+impl VariantCount for u8 {
+	const VARIANT_COUNT: u32 = 256;
+}
```
Comment (ggwpez) [inline substrate/frame/support/src/traits/misc.rs:53]: I think this change itself is fine. If people opt to (ab)use `u8` for Hold or Freeze reason is pretty independant of this.

**Review by ggwpez — APPROVED** [review]

**Comment by bkchr** [comment]
We have `RuntimeHoldReason` and `RuntimeFreezeReason`, why do people need to put there `u8`?

**Comment by kianenigma** [comment]
> We have `RuntimeHoldReason` and `RuntimeFreezeReason`, why do people need to put there `u8`?

Simpler code for starter templates and so on. Imagine how many fewer mental barriers you have to "get your first hold working", if you provide a template to students that works with a constrained `Reason = u8`, rather than needing to pull into `RuntimeHoldReason`, know about the relation between the two, and how to convert from one to other. 

Basically, this tricks prevents one from needing to understand this whole page. 

https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/frame_runtime_types/index.html

None of this is an issue for us and those who are already familiar ofc.

**Comment by bkchr** [comment]
> Imagine how many fewer mental barriers you have to "get your first hold working", if you provide a template to students that works with a constrained `Reason = u8`, rather than needing to pull into `RuntimeHoldReason`, know about the relation between the two, and how to convert from one to other.

Not sure how providing a random `u8` is easier to understand. You are injecting the other runtime types as well using the `derive_impl` macro, not sure why you can not do the same for these types?

https://github.com/paritytech/polkadot-sdk/blob/d8c2944741b1f1b43fa37951fb311d363c753e4e/substrate/frame/nomination-pools/src/lib.rs#L3464 having an enum for hold/freeze is IMO much cleaner. It is like you try to campaign for using C style here, which would be an u8.  [Here](https://github.com/kianenigma/flite/blob/420feba548bc3f1d5ded2cf42718d60df1861833/simple-pallet/src/lib.rs#L82) you just have a `0` as reason. I would be much more confused on what this `0` means and what I should use there in a real application. Or should I use `0` everywhere? You will in the end also write more docs instead of giving the thing a fixed name `WHATEVER_HOLD` and it explains the user right away what it is about.

**Comment by bkchr** [comment]
But if you really think that a plain `0` is better there, I will approve this.

**Review by bkchr — APPROVED** [review]

**Pull request merged**